Introducing new component - Deep Search to show relevant internal documents in addition to results from Web

vijaysaayi opened the pull request:
Deep Search shows relevant internal documents  along with the results from the Web.

- Updated the user interface to show results from Deep search. 
- By default, the results from CSS Wiki are shown. 
- A checkbox is provided to toggle to get results from the CSS Wiki and web.
- Applens App Settings define whether this component is enabled or not

**Sneak-Peek:**
![image](https://user-images.githubusercontent.com/4495262/93083377-20038500-f6b0-11ea-81e0-f44c5eaa1721.png)

SharmaAjay19 commented on AngularApp/projects/diagnostic-data/src/lib/components/documents-search/documents-search.component.ts:
```diff
@@ -0,0 +1,252 @@
+          
+            this.searchTermDisplay = this.searchTerm.valueOf();
+            this.showSearchTermPractices = true;
+            console.log("showing best practices")
```
Let's remove any console.log statements.

vijaysaayi replied:
I've removed console.log statements

SharmaAjay19 commented on AngularApp/projects/diagnostic-data/src/lib/components/documents-search/documents-search.component.ts:
```diff
@@ -0,0 +1,252 @@
+          }));
+    
+    this.showPreLoader = true;
+    console.log(this.searchId);
```
Let's remove any console.log statements.
Just do a quick search for console.log statements in the project folder and if any is introduced in this PR, let's remove them.

vijaysaayi replied:
I've removed console.log statements

SharmaAjay19 commented on ApplensBackend/appsettings.json:
```diff
@@ -97,6 +97,11 @@
+  "DeepSearch": {
+    "endpoint" : "",
+    "authKey": "",
+    "isEnabled" : ""
```
Lets keep isEnabled as false by default in appsettings since the settings are not added yet.

vijaysaayi replied:
I've set the default value of isEnabled as false.

SharmaAjay19 commented on Backend/appsettings.json:
```diff
@@ -25,6 +25,11 @@
+  "DeepSearch": {
+    "endpoint" : "",
+    "authKey": "",
+    "isEnabled" : "true"
```
Lets keep isEnabled as false by default in appsettings since the settings are not added yet.

vijaysaayi replied:
I've set the default value of isEnabled as false

SharmaAjay19 commented:
We should try to limit the height of each article box by limiting the length of the content.
![image](https://user-images.githubusercontent.com/8492235/94206977-84c29a80-fe7b-11ea-915d-b1812f6b2b0e.png)

Here we can see that due to larger content in second article, the boxes have become uneven.

SharmaAjay19 commented:
The minimization of web-search component should occur only when document-search component is embedded. So the view/hide remaining articles should be based on whether document-search is enabled or not. If document-search is disabled, we should display full view of web-search component without the view/hide thing.

vijaysaayi commented:
> We should try to limit the height of each article box by limiting the length of the content.
> ![image](https://user-images.githubusercontent.com/8492235/94206977-84c29a80-fe7b-11ea-915d-b1812f6b2b0e.png)
> 
> Here we can see that due to larger content in second article, the boxes have become uneven.

Fixed the height of article box to 120px and limited the length of description to 300 characters.

> The minimization of web-search component should occur only when document-search component is embedded. So the view/hide remaining articles should be based on whether document-search is enabled or not. If document-search is disabled, we should display full view of web-search component without the view/hide thing.

Introduced a method in web-search to check deep-search enabled status and updated the condition for collapsibility accordingly.

SharmaAjay19 commented on AngularApp/projects/diagnostic-data/src/lib/components/documents-search/documents-search.component.scss:
```diff
@@ -45,10 +45,12 @@ a{
     padding: 10px;
     background-color: white;
     margin: 10px 0px;
+    height: 120px;        
```
Let's keep a max-height rather than keeping a fixed height. When content is smaller than the box, it will help avoid any extra whitespace.

vijaysaayi replied:
updated this to max-height

SharmaAjay19 commented on AngularApp/projects/diagnostic-data/src/lib/components/web-search/web-search.component.ts:
```diff
@@ -165,5 +169,37 @@ export class WebSearchComponent extends DataRenderBaseComponent implements OnIni
+                                    }));
+                                }), 
+                                catchError(e => {
+                                    this.handleRequestFailure();
```
Since this is not an error with web-search itself, we should probably not treat it like a web-search failure and handle it differently.

vijaysaayi replied:
When any errors are encountered, setting the value of deepSearchEnabled as false.

SharmaAjay19 commented on AngularApp/projects/diagnostic-data/src/lib/components/web-search/web-search.component.ts:
```diff
@@ -165,5 +169,37 @@ export class WebSearchComponent extends DataRenderBaseComponent implements OnIni
+        (err) => {  this.logEvent(TelemetryEventNames.DeepSearchResults, { 
+                    operationStatus: "failed",
+                    error: JSON.stringify(err), 
+                    ts: Math.floor((new Date()).getTime() / 1000).toString() 
```
We have to be careful here. If there is an error in determining whether the deep search is enabled, we should make the assumption that deep search is not enabled and render the web-search component accordingly.
Also we probably don't need to log this.

vijaysaayi replied:
When any error is encountered, setting the value of deepSearchEnabled as false, removed logging to this error.

SharmaAjay19 commented on AngularApp/projects/diagnostic-data/src/lib/components/web-search/web-search.component.ts:
```diff
@@ -58,6 +61,7 @@ export class WebSearchComponent extends DataRenderBaseComponent implements OnIni
     }
 
     refresh() {
+        this.checkIfDeepSearchIsEnabled ()
```
Maybe do this in the constructor or OnInit method.

vijaysaayi replied:
Moved the method call to constructor.
